### PR TITLE
Ecal data recovery

### PR DESCRIPTION
**Pull request by argiro**
Ad-hoc recovery sequence for corrupted Ecal headers in run  274157

### CODE REVIEW

File: EventFilter/EcalRawToDigi/plugins/EcalRawDataRecovery.cc [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  //sanity checks:
+  if(rawData_->FEDData(619).size() < minDataSize) return;
+  if(rawData_->FEDData(620).size() < minDataSize) return;
+
+  const uint64_t * pSrpDataFed619 = reinterpret_cast<const uint64_t*>(rawData_->FEDData(619).data() + srpBlockOffset);
+  const uint64_t * pSrpDataFed620 = reinterpret_cast<const uint64_t*>(rawData_->FEDData(620).data() + srpBlockOffset);
+  
+  fixedSrpHeader = pSrpDataFed619[0];
+  //update the SRP-DCC channel value for FED 620 (SRP CH: 5):
+  fixedSrpHeader = (fixedSrpHeader & 0xFFFFFFFFFFFFFF0FLL) | 0x50;
```
Comment (slava77) [inline EventFilter/EcalRawToDigi/plugins/EcalRawDataRecovery.cc]: can't this part be done in the unpacker itself?

File: EventFilter/EcalRawToDigi/interface/EcalRawDataRecovery.h [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+class FEDRawDataCollection;
+
+/** EcalRawDataRecover class
+ *
+ * CMSSW module to fix identified ECAL data corruption. Current implementation
+ * fixes the problem of run 274157 (see fixRun274157). The code can be extended
+ * to deal with data issue that would be found in new runs.
+ * Original author: Ph. Gras CEA/Saclay
+ */
+class EcalRawDataRecovery : public edm::EDProducer {
```
Comment (slava77) [inline EventFilter/EcalRawToDigi/interface/EcalRawDataRecovery.h]: new modules have to be MT-aware https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkModuleTypes

**Comment by slava77** [comment]
I'd really prefer the recovery is done in full inside the unpacker,
without extra modules.
What prevents this problem from happening again? (in that case "if run ==x" will not work)

**Comment by argiro** [comment]
adding @grasph to answer questions about the code

File: Configuration/StandardSequences/python/RawToDigi_Data_cff.py [hunk 3]
```diff
@@ -2,10 +2,12 @@
 
 from Configuration.StandardSequences.RawToDigi_cff import *
 
+from EventFilter.EcalRawToDigi.EcalRawDataRecovery_cfi import ecalRawDataRecovery
+
 RawToDigi = cms.Sequence(
                          siPixelDigis
                          +siStripDigis
-                         +ecalDigis
+                         +(ecalRawDataRecovery*ecalDigis)
```
Comment (slava77) [inline Configuration/StandardSequences/python/RawToDigi_Data_cff.py]: modifications in this file will need to be made in a different way in CMSSW_8_0_9 and later

File: EventFilter/EcalRawToDigi/plugins/EcalRawDataRecovery.cc [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  nOkEvents_(0)
+{
+  produces<FEDRawDataCollection>(ecalRawDataOut_);
+}
+
+void EcalRawDataRecovery::beginJob(){
+}
+
+void EcalRawDataRecovery::endJob(){
+  /*edm::LogInfo("EcalRawDataRecovery")*/ std::cout << "Number of processed events: " << iEvent_
```
Comment (slava77) [inline EventFilter/EcalRawToDigi/plugins/EcalRawDataRecovery.cc]: cout is not allowed in production code.

**Comment by grasph** [comment]
I can move the code into the unpacker module.  It will still be a preprocessing of the raw data, as I don't have enough understanding of unpacker code to embed it in the unpacking process. It will fix at same time the multithreading issue, which I've not understood, and the configuration change for CMSSW 8.0.9 and higher.

The data were corrupted due to a hardware issue in a FED. It's the first time it happens, so we have little experience with it (and I hope we won't have more). The fix was designed by looking at the raw data of this run and it's specific to these data.

Philippe.

**Comment by slava77** [comment]
On 5/30/16 1:09 PM, grasph wrote:
> I can move the code into the unpacker module. It will still be a
> preprocessing of the raw data, as I don't have enough understanding of
> unpacker code to embed it in the unpacking process.

Does this mean a part of the text that makes
the recoveredRawData in situ inside  EcalRawToDigi::produce
and will pass it otherwise the same to the DCCDataUnpacker::unpack ?
I think it's probably OK.
Eventually, it makes more sense to have this handled appropriately
inside DCCDataUnpacker code


> It will fix at same
> time the multithreading issue, which I've not understood, and the
> configuration change for CMSSW 8.0.9 and higher.
> 
> The data were corrupted due to a hardware issue in a FED. It's the first
> time it happens, so we have little experience with it (and I hope we
> won't have more). The fix was designed by looking at the raw data of
> this run and it's specific to these data.

Unless you know which hardware failed and why and that it was fixed already,
I think you need to be prepared to handle this corruption case in general.


Slava

**Comment by grasph** [comment]
We agree with your last statement. It is planned to make a longer term 
fix. If the rereco is in few weeks, then we may not need anymore this 
quick fix. I need to discuss with Stefano and ECAL colleagues on the 
plan, We will come back to you.

Thanks for the code review,

Philippe.

On 30/05/16 22:57, Slava Krutelyov wrote:
> On 5/30/16 1:09 PM, grasph wrote:
>> I can move the code into the unpacker module. It will still be a
>> preprocessing of the raw data, as I don't have enough understanding of
>> unpacker code to embed it in the unpacking process.
>
> Does this mean a part of the text that makes
> the recoveredRawData in situ inside EcalRawToDigi::produce
> and will pass it otherwise the same to the DCCDataUnpacker::unpack ?
> I think it's probably OK.
> Eventually, it makes more sense to have this handled appropriately
> inside DCCDataUnpacker code
>
>
>> It will fix at same
>> time the multithreading issue, which I've not understood, and the
>> configuration change for CMSSW 8.0.9 and higher.
>>
>> The data were corrupted due to a hardware issue in a FED. It's the first
>> time it happens, so we have little experience with it (and I hope we
>> won't have more). The fix was designed by looking at the raw data of
>> this run and it's specific to these data.
>
> Unless you know which hardware failed and why and that it was fixed already,
> I think you need to be prepared to handle this corruption case in general.
>
>
> Slava
>
>
>>
>> Philippe.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/cms-sw/cmssw/pull/14689#issuecomment-222548771>, or
>> mute the thread
>> <https://github.com/notifications/unsubscribe/AEdcbizwEvTPt9dsxvSE3JsATGFPCFRFks5qG0QUgaJpZM4Ip_GH>.
>>
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/14689#issuecomment-222553697>, or
> mute the thread
> <https://github.com/notifications/unsubscribe/AFHdXiPOXLwTIxsvvVzsQAqvgnA9WAwUks5qG09QgaJpZM4Ip_GH>.
>

-- 
Philippe GRAS                      CEA Saclay. IRFU/SEDI
Phone: +33 1 69 08 62 81           F-91191 Gif sur Yvette Cedex - FRANCE
Fax:   +33 1 69 08 31 47           Office 141-111
.............................................................................
CERN Office: 42-1-024              Tel : +41 22 766 2389
=============================================================================

**Comment by argiro** [comment]
Closing this PR as we want to go for a different strategy

**Pull request closed without merging**